/*
 * Copyright (c) 2012 - 2025 Data In Motion and others.
 * All rights reserved. 
 * 
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 * 
 * SPDX-License-Identifier: EPL-2.0
 * 
 * Contributors:
 *      Mark Hoffmann - initial API and implementation
 */
package org.cyclonedx.schema.bom.impl;

import java.util.Collection;

import org.cyclonedx.schema.bom.BOMPackage;
import org.cyclonedx.schema.bom.EnergyConsumptionsType;
import org.cyclonedx.schema.bom.EnvironmentalConsiderationsType;
import org.cyclonedx.schema.bom.PropertiesType;

import org.eclipse.emf.common.notify.Notification;
import org.eclipse.emf.common.notify.NotificationChain;

import org.eclipse.emf.common.util.EList;

import org.eclipse.emf.ecore.EClass;
import org.eclipse.emf.ecore.InternalEObject;

import org.eclipse.emf.ecore.impl.ENotificationImpl;
import org.eclipse.emf.ecore.impl.MinimalEObjectImpl;

import org.eclipse.emf.ecore.util.EObjectContainmentEList;
import org.eclipse.emf.ecore.util.InternalEList;

/**
 * <!-- begin-user-doc -->
 * An implementation of the model object '<em><b>Environmental Considerations Type</b></em>'.
 * <!-- end-user-doc -->
 * <p>
 * The following features are implemented:
 * </p>
 * <ul>
 *   <li>{@link org.cyclonedx.schema.bom.impl.EnvironmentalConsiderationsTypeImpl#getEnergyConsumptions <em>Energy Consumptions</em>}</li>
 *   <li>{@link org.cyclonedx.schema.bom.impl.EnvironmentalConsiderationsTypeImpl#getProperties <em>Properties</em>}</li>
 * </ul>
 *
 * @generated
 */
public class EnvironmentalConsiderationsTypeImpl extends MinimalEObjectImpl.Container implements EnvironmentalConsiderationsType {
	/**
	 * The cached value of the '{@link #getEnergyConsumptions() <em>Energy Consumptions</em>}' containment reference list.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @see #getEnergyConsumptions()
	 * @generated
	 * @ordered
	 */
	protected EList<EnergyConsumptionsType> energyConsumptions;

	/**
	 * The cached value of the '{@link #getProperties() <em>Properties</em>}' containment reference.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @see #getProperties()
	 * @generated
	 * @ordered
	 */
	protected PropertiesType properties;

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	protected EnvironmentalConsiderationsTypeImpl() {
		super();
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	@Override
	protected EClass eStaticClass() {
		return BOMPackage.eINSTANCE.getEnvironmentalConsiderationsType();
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	@Override
	public EList<EnergyConsumptionsType> getEnergyConsumptions() {
		if (energyConsumptions == null) {
			energyConsumptions = new EObjectContainmentEList<EnergyConsumptionsType>(EnergyConsumptionsType.class, this, BOMPackage.ENVIRONMENTAL_CONSIDERATIONS_TYPE__ENERGY_CONSUMPTIONS);
		}
		return energyConsumptions;
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	@Override
	public PropertiesType getProperties() {
		return properties;
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	public NotificationChain basicSetProperties(PropertiesType newProperties, NotificationChain msgs) {
		PropertiesType oldProperties = properties;
		properties = newProperties;
		if (eNotificationRequired()) {
			ENotificationImpl notification = new ENotificationImpl(this, Notification.SET, BOMPackage.ENVIRONMENTAL_CONSIDERATIONS_TYPE__PROPERTIES, oldProperties, newProperties);
			if (msgs == null) msgs = notification; else msgs.add(notification);
		}
		return msgs;
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	@Override
	public void setProperties(PropertiesType newProperties) {
		if (newProperties != properties) {
			NotificationChain msgs = null;
			if (properties != null)
				msgs = ((InternalEObject)properties).eInverseRemove(this, EOPPOSITE_FEATURE_BASE - BOMPackage.ENVIRONMENTAL_CONSIDERATIONS_TYPE__PROPERTIES, null, msgs);
			if (newProperties != null)
				msgs = ((InternalEObject)newProperties).eInverseAdd(this, EOPPOSITE_FEATURE_BASE - BOMPackage.ENVIRONMENTAL_CONSIDERATIONS_TYPE__PROPERTIES, null, msgs);
			msgs = basicSetProperties(newProperties, msgs);
			if (msgs != null) msgs.dispatch();
		}
		else if (eNotificationRequired())
			eNotify(new ENotificationImpl(this, Notification.SET, BOMPackage.ENVIRONMENTAL_CONSIDERATIONS_TYPE__PROPERTIES, newProperties, newProperties));
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	@Override
	public NotificationChain eInverseRemove(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
		switch (featureID) {
			case BOMPackage.ENVIRONMENTAL_CONSIDERATIONS_TYPE__ENERGY_CONSUMPTIONS:
				return ((InternalEList<?>)getEnergyConsumptions()).basicRemove(otherEnd, msgs);
			case BOMPackage.ENVIRONMENTAL_CONSIDERATIONS_TYPE__PROPERTIES:
				return basicSetProperties(null, msgs);
		}
		return super.eInverseRemove(otherEnd, featureID, msgs);
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	@Override
	public Object eGet(int featureID, boolean resolve, boolean coreType) {
		switch (featureID) {
			case BOMPackage.ENVIRONMENTAL_CONSIDERATIONS_TYPE__ENERGY_CONSUMPTIONS:
				return getEnergyConsumptions();
			case BOMPackage.ENVIRONMENTAL_CONSIDERATIONS_TYPE__PROPERTIES:
				return getProperties();
		}
		return super.eGet(featureID, resolve, coreType);
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	@SuppressWarnings("unchecked")
	@Override
	public void eSet(int featureID, Object newValue) {
		switch (featureID) {
			case BOMPackage.ENVIRONMENTAL_CONSIDERATIONS_TYPE__ENERGY_CONSUMPTIONS:
				getEnergyConsumptions().clear();
				getEnergyConsumptions().addAll((Collection<? extends EnergyConsumptionsType>)newValue);
				return;
			case BOMPackage.ENVIRONMENTAL_CONSIDERATIONS_TYPE__PROPERTIES:
				setProperties((PropertiesType)newValue);
				return;
		}
		super.eSet(featureID, newValue);
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	@Override
	public void eUnset(int featureID) {
		switch (featureID) {
			case BOMPackage.ENVIRONMENTAL_CONSIDERATIONS_TYPE__ENERGY_CONSUMPTIONS:
				getEnergyConsumptions().clear();
				return;
			case BOMPackage.ENVIRONMENTAL_CONSIDERATIONS_TYPE__PROPERTIES:
				setProperties((PropertiesType)null);
				return;
		}
		super.eUnset(featureID);
	}

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @generated
	 */
	@Override
	public boolean eIsSet(int featureID) {
		switch (featureID) {
			case BOMPackage.ENVIRONMENTAL_CONSIDERATIONS_TYPE__ENERGY_CONSUMPTIONS:
				return energyConsumptions != null && !energyConsumptions.isEmpty();
			case BOMPackage.ENVIRONMENTAL_CONSIDERATIONS_TYPE__PROPERTIES:
				return properties != null;
		}
		return super.eIsSet(featureID);
	}

} //EnvironmentalConsiderationsTypeImpl
